### PR TITLE
Add support for timeout on getting the next ssl record for a connection

### PR DESCRIPTION
**Pull request by npmenard**

### CODE REVIEW

File: micro-rdk/src/esp32/dtls.rs [hunk 1]
```diff
@@ -107,14 +91,36 @@ unsafe extern "C" fn mbedtls_net_read<S: Read>(
...
+    let buf = std::slice::from_raw_parts_mut(buf as *mut _, len);
+    match state.read(buf) {
+        Ok(len) => {
+            // if operation succeed cancel running timeout
+            state.reset_read_timeout();
+            len as c_int
+        }
+        Err(e) => {
+            let _ = state.error.insert(e);
+            if state.error.as_ref().unwrap().kind() == std::io::ErrorKind::WouldBlock {
```
Comment (gvaradarajan) [inline micro-rdk/src/esp32/dtls.rs:115]: match statement instead?

File: micro-rdk/src/esp32/dtls.rs [hunk 2]
```diff
@@ -139,37 +145,114 @@ extern "C" fn ssl_debug(
...
+    fn get_delay(&self) -> i32 {
+        if self.fin.is_none() {
+            return -1;
+        }
+        let now = Instant::now();
+        if now > *self.intermediate.as_ref().unwrap() {
+            log::debug!("intermetidate timer expired");
+            return 2;
+        }
+        if now > *self.fin.as_ref().unwrap() {
```
Comment (gvaradarajan) [inline micro-rdk/src/esp32/dtls.rs:211]: shouldn't we check fin before intermediate?

File: micro-rdk/src/esp32/dtls.rs [hunk 3]
```diff
@@ -426,24 +510,24 @@ impl SSLContext {
     }
 }
 
-pub struct SSLStream<S> {
+pub(crate) struct SSLStream<S> {
```
Comment (gvaradarajan) [inline micro-rdk/src/esp32/dtls.rs:513]: this is holding a pointer and interacting with it in an unsafe way, is it ok that you don't have a drop implementation for de-allocating the memory behind the pointer?

**Review by gvaradarajan — CHANGES_REQUESTED** [review]
Had a couple questions

File: micro-rdk/src/esp32/dtls.rs (hunk 1 above)
Comment (npmenard) [reply micro-rdk/src/esp32/dtls.rs]: good point

File: micro-rdk/src/esp32/dtls.rs (hunk 2 above)
Comment (npmenard) [reply micro-rdk/src/esp32/dtls.rs:209]: nice catch that was really wrong :)

File: micro-rdk/src/esp32/dtls.rs (hunk 3 above)
Comment (npmenard) [reply micro-rdk/src/esp32/dtls.rs:511]: if you are talking about bio_ptr it should be dropped on line 519

Comment (gvaradarajan) [reply micro-rdk/src/esp32/dtls.rs:511]: Ah I missed that, my bad

**Review by gvaradarajan — APPROVED** [review]
LGTM

**Pull request merged**